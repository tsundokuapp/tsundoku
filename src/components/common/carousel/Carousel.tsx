'use client';
// Color Checked
import {
  ArrowCircleLeft,
  ArrowCircleRight,
} from '@phosphor-icons/react/dist/ssr';
import Image from 'next/image';
import { useEffect, useState } from 'react';

import { cn } from '@/helpers/twUtils';

interface CarouselProps {
  images: string[];
  autoSlide?: boolean;
  autoSlideInterval?: number;
  navigation?: boolean;
  className?: string;
}

export const Carousel = ({
  images,
  autoSlide = true,
  autoSlideInterval = 3000,
  navigation = false,
}: CarouselProps) => {
  const [currentIndex, setCurrentIndex] = useState(0);

  useEffect(() => {
    if (autoSlide) {
      const slideInterval = setInterval(() => {
        setCurrentIndex((prevIndex) => (prevIndex + 1) % images.length);
      }, autoSlideInterval);
      return () => clearInterval(slideInterval);
    }
  }, [autoSlide, autoSlideInterval, images.length]);

  const nextSlide = () => {
    setCurrentIndex((prevIndex) => (prevIndex + 1) % images.length);
  };

  const prevSlide = () => {
    setCurrentIndex(
      (prevIndex) => (prevIndex - 1 + images.length) % images.length,
    );
  };
  return (
    <div className="relative w-full">
      <div className="relative h-72 w-full overflow-hidden rounded-md sm:h-64 sm:rounded-3xl">
        {images.map((image, index) => (
          <div
            key={index}
            className={cn(
              `absolute inset-0 w-full translate-x-full transform transition-transform duration-500`,
              {
                'translate-x-0': index === currentIndex,
              },
            )}
          >
            <Image
              height={300}
              width={350}
              src={image}
              alt={`Slide ${index}`}
              className="h-full w-full object-cover"
            />
          </div>
        ))}
      </div>
      {navigation && (
        <>
          <button
            className="absolute left-0 top-1/2 -translate-y-1/2 transform p-2 text-appButtonText"
            onClick={prevSlide}
            aria-label="previous"
            data-testid="previous-button"
          >
            <ArrowCircleLeft size={24} />
          </button>
          <button
            className="absolute right-0 top-1/2 -translate-y-1/2 transform p-2 text-appButtonText"
            onClick={nextSlide}
            aria-label="next"
            data-testid="next-button"
          >
            <ArrowCircleRight size={24} />
          </button>
        </>
      )}
    </div>
  );
};
