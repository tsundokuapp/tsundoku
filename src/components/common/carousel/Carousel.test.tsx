import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { Carousel } from './Carousel';

describe('Testes para o componente <Carousel />', () => {
  describe('Testes de renderização', () => {
    it('Deve renderizar Carousel corretamente', () => {
      // Given
      render(
        <Carousel images={['/img1.jpg', '/img2.jpg']} autoSlide={false} />,
      );

      // When
      const firstImage = screen.getByAltText('Slide 0');

      // Then
      expect(firstImage).toBeInTheDocument();
    });

    it('Deve exibir a primeira imagem ao iniciar', () => {
      // Given
      render(
        <Carousel images={['/img1.jpg', '/img2.jpg']} autoSlide={false} />,
      );

      // When
      const firstImage = screen.getByAltText('Slide 0');

      // Then
      expect(firstImage).toHaveAttribute(
        'src',
        expect.stringMatching(/\/_next\/image\?url=%2Fimg1.jpg&/),
      );
    });

    it('Deve renderizar os botões quando a prop "navigation" for true', async () => {
      // Given
      render(
        <Carousel
          images={['/img1.jpg', '/img2.jpg']}
          autoSlide={false}
          navigation
        />,
      );

      // When
      const nextButton = screen.getByTestId('next-button');
      const prevButton = screen.getByTestId('previous-button');

      // Then
      expect(nextButton).toBeInTheDocument();
      expect(prevButton).toBeInTheDocument();
    });

    it('Não deve quebrar se a lista de imagens estiver vazia', () => {
      // Given
      render(<Carousel images={[]} />);

      // Then
      expect(screen.queryByRole('img')).not.toBeInTheDocument();
    });
  });

  describe('Testes de interação e comportamento', () => {
    jest.useFakeTimers();

    it('Deve avançar para a próxima imagem ao clicar no botão "próximo"', async () => {
      // Given
      render(
        <Carousel
          images={['/img1.jpg', '/img2.jpg']}
          autoSlide={false}
          navigation
        />,
      );

      // When
      const nextButton = screen.getByTestId('next-button');
      fireEvent.click(nextButton);

      // Then
      expect(screen.getByAltText('Slide 1')).toBeInTheDocument();
    });

    it('Deve voltar para a imagem anterior ao clicar no botão "anterior"', async () => {
      // Given
      render(
        <Carousel
          images={['/img1.jpg', '/img2.jpg']}
          autoSlide={false}
          navigation
        />,
      );

      // When
      const prevButton = screen.getByTestId('previous-button');
      fireEvent.click(prevButton);

      // Then
      expect(screen.getByAltText('Slide 1')).toBeInTheDocument(); // Última imagem, pois circula para trás
    });

    it('Deve avançar automaticamente se autoSlide for true', () => {
      // Given
      render(
        <Carousel
          images={['/img1.jpg', '/img2.jpg']}
          autoSlide={true}
          autoSlideInterval={3000}
        />,
      );

      // When
      jest.advanceTimersByTime(3000);

      // Then
      expect(screen.getByAltText('Slide 1')).toBeInTheDocument();
    });

    it('Não deve avançar automaticamente se autoSlide for false', () => {
      // Given
      render(
        <Carousel images={['/img1.jpg', '/img2.jpg']} autoSlide={false} />,
      );

      // When
      jest.advanceTimersByTime(3000);

      // Then
      expect(screen.getByAltText('Slide 0')).toBeInTheDocument();
    });
  });

  describe('Teste de snapshot', () => {
    it('Deve manter o snapshot do Carousel', () => {
      // Given
      const { container } = render(
        <Carousel images={['/img1.jpg', '/img2.jpg']} />,
      );

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
