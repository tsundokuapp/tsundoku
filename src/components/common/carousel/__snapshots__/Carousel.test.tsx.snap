// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes para o componente <Carousel /> Teste de snapshot Deve manter o snapshot do Carousel 1`] = `
<div>
  <div
    class="relative w-full max-w-3xl"
  >
    <div
      class="relative h-72 w-full overflow-hidden rounded-md"
    >
      <div
        class="absolute inset-0 w-full transform transition-transform duration-500 translate-x-0"
      >
        <img
          alt="Slide 0"
          class="h-full w-full object-cover"
          data-nimg="1"
          decoding="async"
          height="300"
          loading="lazy"
          src="/_next/image?url=%2Fimg1.jpg&w=750&q=75"
          srcset="/_next/image?url=%2Fimg1.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg1.jpg&w=750&q=75 2x"
          style="color: transparent;"
          width="350"
        />
      </div>
      <div
        class="absolute inset-0 w-full translate-x-full transform transition-transform duration-500"
      >
        <img
          alt="Slide 1"
          class="h-full w-full object-cover"
          data-nimg="1"
          decoding="async"
          height="300"
          loading="lazy"
          src="/_next/image?url=%2Fimg2.jpg&w=750&q=75"
          srcset="/_next/image?url=%2Fimg2.jpg&w=384&q=75 1x, /_next/image?url=%2Fimg2.jpg&w=750&q=75 2x"
          style="color: transparent;"
          width="350"
        />
      </div>
    </div>
    <button
      aria-label="previous"
      class="absolute left-0 top-1/2 -translate-y-1/2 transform p-2 text-white"
      data-testid="previous-button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm48-88a8,8,0,0,1-8,8H107.31l18.35,18.34a8,8,0,0,1-11.32,11.32l-32-32a8,8,0,0,1,0-11.32l32-32a8,8,0,0,1,11.32,11.32L107.31,120H168A8,8,0,0,1,176,128Z"
        />
      </svg>
    </button>
    <button
      aria-label="next"
      class="absolute right-0 top-1/2 -translate-y-1/2 transform p-2 text-white"
      data-testid="next-button"
    >
      <svg
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm45.66-93.66a8,8,0,0,1,0,11.32l-32,32a8,8,0,0,1-11.32-11.32L148.69,136H88a8,8,0,0,1,0-16h60.69l-18.35-18.34a8,8,0,0,1,11.32-11.32Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
